server/activate.go

```diff
@@ -24,89 +23,105 @@ func (p *Plugin) OnActivate() error {
 		p.serverVersion = serverVersion
 	}
 
-	if err := p.ensureBotExists(); err != nil {
+	if botUserId, err := p.ensureBotExists(); err != nil {
 		return errors.Wrap(err, "failed to ensure bot user exists")
+	} else {
+		p.botUserId = botUserId
 	}
 
 	p.initializeClient()
 
 	p.registerCommands()
 
+	p.setActivated(true)
 	p.API.LogDebug("NPS plugin activated")
 
-	p.checkForNextSurvey(p.serverVersion)
+	go p.checkForNextSurvey(p.serverVersion)
 
 	return nil
 }
 
+func (p *Plugin) setActivated(activated bool) {
+	p.activatedLock.Lock()
+	defer p.activatedLock.Unlock()
+
+	p.activated = activated
+}
+
+func (p *Plugin) isActivated() bool {
+	p.activatedLock.RLock()
+	defer p.activatedLock.RUnlock()
+
+	return p.activated
+}
+
 func (p *Plugin) canSendDiagnostics() bool {
 	enableDiagnostics := p.API.GetConfig().LogSettings.EnableDiagnostics
 	return enableDiagnostics != nil && *enableDiagnostics
 }
 
-func (p *Plugin) ensureBotExists() *model.AppError {
+func (p *Plugin) ensureBotExists() (string, *model.AppError) {
 	// Attempt to find an existing bot
 	var botUserId string
-	appErr := p.KVGet(BOT_USER_KEY, &botUserId)
-	if appErr != nil {
-		return appErr
+	if appErr := p.KVGet(BOT_USER_KEY, &botUserId); appErr != nil {
```

Comment:
Can't we rely on `CreateBot` to fail if the bot already exists, negating the need to store this in the KVStore and thus eliminate the race?

Reply:
Yeah, that's true. It's slightly more expensive to attempt to create the user and then find them, but it's a negligible increase in run time to reduce the complexity of the code. I'll look at changing that